### ISSUE TITLE
Zaptec Go: Authorising a charge service

### ISSUE THREAD

**Issue by JensDebergh** [custom-components/zaptec#55]
Pretty new to home assistant and tried the integration in an automation flow.

Is the start charging call with authorization enabled failing for anyone else? The stop charging works perfectly fine given a proper `charger_id`.

I'm able to pause and restart, but not start from the beginning. Am I forgetting something or is the call broken currently?

command: `501` doesn't seem to be listed anymore on https://api.zaptec.com/help/index.html#/Charger/post_api_chargers__id__sendCommand__commandId_`

**Comment by sveinse** [comment]
What version of the integration are you using?

I'm not completely following your flow, but I believe the service to use to start charging is `resume_charging` which issues the `507`command.

**Comment by JensDebergh** [comment]
I have a zaptec go hooked up to home assistant.

What I want to achieve is to trigger an automation that runs at 10PM to authorise the charger to start charging at that time.

I tried the "start charging" service, but it doesn't seem to work. Do I use the "resume_charging" one? Even if I'm not authorized yet?

**Comment by sveinse** [comment]
What do you mean by "authorize"? Authorized as in a security means or authorize as in wanting to have a delayed start?

Do you have any security enabled on the charger at all, e.g. with RFID?

**Comment by JensDebergh** [comment]
Apologies for not being clear enough.

The zaptec is locked by default so plugging in a cable is not going to auto charge it.

I need to "authorise" by going to the zaptec app and click the "authorise" button to allow the zaptec charger to charge.

I want home assistant to signal: hey it's me you can start charging.

I was under the impression I could use the start_charging service for this.

**Comment by sveinse** [comment]
Got it. I'm not sure how or if authorization from HA is possible. I need to check the code and Zaptec integration docs tomorrow.

Say this isn't possible, would the opposite procedure be an alternative: Connect the car, authorize with RFID on the charger, leave. HA will then start the charging on timed automation. The cable remains locked until the charge is done.

I'm not at all sure if this is possible  -- I need to experiment with it.

**Comment by sveinse** [comment]
It is very interesting to get these inputs, and it's very useful to understand what the (wanted) user flow is. I've learned that there are many different ways users connect, use and automate EV charging. Getting this input is very useful for the continued development of it.

**Comment by JensDebergh** [comment]
Hi @sveinse 

My current setup charging my car goes as followed.

1) I live in Belgium and we have 2 different rates based on the time. Day rate (7am - 10pm) and night rate (10pm - 7am)

2) For obvious reason I charge my car at night since the rate is almost half the day rate.

3) I do not own a garage where I can leave my zaptec charger unauthorized

4) For this reason I have an authorized setup where after plugging in the cable to my car I need to authorise the charging sessions.  From a quick glance at the code this is what missing for me to automate charging starting from 10pm by home assistant.

---

At 10pm I would like to automate home assistant to authorise the charge if it hasn't already been enabled. This can be done by looking at several sensors from the car & charger.

---

I did some investigating myself and put up a reverse proxy between the mobile app from zaptec and my iphone.

This is an undocumented API call that is not inside their official documentation, but you can just send in the required information below. I've tested this numerous times via Curl commands.

The API call that needs to be added is:

Method: `POST`
URL: `https://api.zaptec.com/api/chargers/<charger_id>/authorizecharge`
Headers:

```
Authorization: Bearer <ACCESS_TOKEN>
```

Maybe we also need:

```
user-agent: Zaptec/5.9.2 (com.zaptec.zapapp; build:1559; iOS 16.3.1; Model:iPhone) CFNetwork/1404.0.5
```

Once authorized the vehicle will start charging and from there on the services already in this repo work like a charm. 

I'm not familiar with python specifically but I was wondering if you could add this call?

**Comment by sveinse** [comment]
According to Zaptec integrator documentation (https://zaptec.com/downloads/ZapChargerPro_Integration.pdf) there are three methods of authentication: Zaptec internal, webhooks and OCPP-J. Only the two latter are described in the docs. What you're describing is most likely a reverse-engineer of their internal app authorization format.

What kind of payload did this authorizationcharge message have?

**Comment by JensDebergh** [comment]
Hi @sveinse 

That's correct. I proxied the requests through my laptop to see what payload is needed to authorise the charge.

If the autorisation is a success you get no payload back and only a http status code `200`

If a session is already running and you call the authorise endpoint you get an error message JSON payload.

I was whipping up a merge request:

Inside of `services.py`

```
    async def service_handle_authorize_charge(service_call):
        _LOGGER.debug("service authorize charge")
        charger_id = service_call.data["charger_id"]
        cmd = f"chargers/{charger_id}/authorizecharge"
        return await acc._request(cmd, method="post")
        
    hass.services.async_register(
        DOMAIN, "authorize_charge", service_handle_authorize_charge, schema=has_id_schema
    )
```

Inside of `services.yaml`

```
# Needs to be added
authorise_charging:
  description: "Authorises a charging session"
  fields:
    charger_id:
      description: "The charger id"
      example: "yyyyyyyy-9999-99xx-xxx-yyyyyyyyyyyy"
```

Because I'm using the existing `acc_request(cmd, method="post")`, which auto parses to `JSON`, I'm getting the following error:

```
Could not get info from https://api.zaptec.com/api/chargers/<CHARGER ID>/authorizecharge: 0, message='Attempt to decode JSON with unexpected mimetype: ', url=URL('https://api.zaptec.com/api/chargers/<CHARGER ID>/authorizecharge')
```

The authorisation call succeeds, but the parsing goes wrong (but it works)

I come from a ruby background and my python knowledge is limited.

**Comment by sveinse** [comment]
It does work! I was successful to authorize charging with this!! -- I'm implementing this as a button in my `zaptec-dev` fork.

**Comment by philsson** [comment]
> It does work! I was successful to authorize charging with this!! -- I'm implementing this as a button in my `zaptec-dev` fork. (edit: here https://github.com/sveinse/zaptec)

Would be nice if other services could also be implemented as buttons. Maybe the charger id sets the limit for this. I'm not read up on the details.

**Comment by sveinse** [comment]
Everything is implemented as buttons in my fork, no services at all. The fork is a major refactor of the whole zaptec component with considerable improvements which I hope the official repo is interested in taking on board.

**Comment by Hellowlol** [comment]
@JensDebergh Yes, im 99,99% sure that the correct way to go would be to call resume_charging service. Can anybody test it?

**Comment by sveinse** [comment]
@Hellowlol is is possible to resume an authorized charging session that have been paused

**Comment by Hellowlol** [comment]
Why do you want to deauthorize the session? If you pause it with stop_pause_charging you should be able to just start it with resume_charging.

**Comment by sveinse** [comment]
Why do users have/want authorization? I think the answer to that question will determine the need for deauthorize or not.

**Comment by Hellowlol** [comment]
I understand why people have chargers that they need "unlock", but I don't understand why people just use  the resume_charging service that handles everything. There shouldn't be a need to call authorize.

**Comment by JensDebergh** [comment]
@Hellowlol Just tested this and the `resume_charging` service does not work (atleast for me with authorisation enabled).

So there is a need for this service.

**Comment by Hellowlol** [comment]
@JensDebergh Thanks for testing Jens! What charger do you have?

**Comment by sveinse** [comment]
Allow me to repeat what I wrote in #49 

> If a car is connected to the charger and it doesn't immediately start charging, it doesn't actually matter if the delayed start of it is done via start charge commands or via authentication from HA. Unless there is a way for HA to verify that the right car is connected, authentication doesn't add anything. A "blind" authenticate command from HA is simply a start charge command without any added security.

**Comment by Hellowlol** [comment]
@JensDebergh Can you also test the service start_charging?

@sveinse What is your point with that statement? How do you plan to know if it's your car or not?, there are many ways to solve this, but I can't think of one using any data we get from zaptec. I assume most users know for sure that it is their car, they parked in front of the charger, put the cable in, the cable is locked between the charger and the car. If the car connected to the users charger isnt really theirs, I suppose they have bigger issues than the 30 NOK they might loose from charging another persons car.

**Comment by sveinse** [comment]
I didn't mean it hostile or disrespect. I'm just trying to understand the wanted intent for enabling authorization. I'm sure you know that authorization and cable locking is not the same thing. If the intent is to lock the cable, one doesn't need authentication to do so.

However, Zaptec unlocks the cable when going into finished (i.e. pause) mode with or without authentication. Hence the only way to reliably secure the charge cable is to use the permanent lock feature.

I have Zaptec Go and I have tested the following:

- Authentication is enabled
- Chargers goes into "waiting" mode on cable insert. The cable is locked when inserted.
- Charging does not start on "start_charging" (501). API returns code 517
- Issuing "stop_charging" (502) does nothing in either mode
- Charging does not start on "resume_charging" (507)
- Sending "authorizecharge" and charging commences, like @JensDebergh has shown
- Issuing "stop_pause_charging" (506) stops charging and puts it into "finished" mode and unlocks the cable. This command can be issued in "waiting" mode as well.
- Charging can be resumed with "resume_charging" (507) if previously authorized
- Sending "deauthorize_stop" (10001) puts it into "disconnected" and then "waiting" mode right after. The cable is unlocked. API returns code 528. From this mode it must be reauthorized to start charging.

**Comment by JensDebergh** [comment]
@sveinse There is also a pause button when you start charging inside of the app. I don't use it often but I can connect my proxy again to see if its using the `sendCommand` endpoint or something like `pausecharging`

@Hellowlol I have a [Zaptec Go](https://www.cablesoolutions.shop/nl/zaptec-go.html?utm_term=&utm_campaign=Smart+Shopping+-+Laadkabels&utm_source=adwords&utm_medium=ppc&hsa_acc=8505721072&hsa_cam=17753779841&hsa_grp=&hsa_ad=&hsa_src=x&hsa_tgt=&hsa_kw=&hsa_mt=&hsa_net=adwords&hsa_ver=3&gclid=EAIaIQobChMIj9L1zOvNgAMVGM53Ch11sQVvEAQYASABEgIcS_D_BwE)

**Comment by sveinse** [comment]
For completeness, I've made similar tests without authentication:

- Authentication is disabled
- Available current is set to 0 to prevent charger to auto start on cable insert.
- Charger goes into "waiting" mode on cable insert. The cable is locked when inserted.
- Charging is started by setting the wanted current (or set it to 32 to let the car limit). It goes into "charging" mode.
- Issuing "stop_pause_charging" (506) stops charging and puts it into "finished" mode and unlocks the cable.
- Issuing "resume_charging" (507) starts charging again (unless current is set to 0).
- Setting available current to 0 sets stop charging and puts it into "waiting" mode and the cable remains locked.

As can be seen here, authentication on/off has no impact on the locking mechanism and thus the use for authentication IMHO mainly to authorize as in identify the user of the charger.

There is one notable difference: By setting available current to 0 in order to pause charging, the cable remains locked. This is the only way I've found that pauses charging and leaves the cable locked. With this method, I'd safely leave my charger and car and let HA figure out when to charge. The cable will remain locked all the time.

**Comment by sveinse** [comment]
> @sveinse There is also a pause button when you start charging inside of the app. I don't use it often but I can connect my proxy again to see if its using the `sendCommand` endpoint or something like `pausecharging`

While I agree implementing the "authorizecharge" function in HA, how wise is it of us to implement these non-documented, reverse engineered features? I'm tempted by the opportunities and worried that it suddenly doesn't work or change behavior...

**Comment by JensDebergh** [comment]
It’s pretty safe because changing API calls takes ages when you have a mobile app.

You need backwards compatibility for older versions of the app.

The alternative is not being able to charge with the authorise flow.

**Comment by sveinse** [comment]
Another use-case is to enable authentication and do authorization by RFID-tags on the unit itself (and not do authentication from HA). The process of authentication and (delayed) start/stop from HA can be implemented as described below.

For the user the flow is as follows:
- Connect the charger
- Authenticate with an RFID tag
- HA will through automation chose when to start and stop the charging
- When wanting to leave, release the lock by presenting the RFID tag once more.

As long as the start and stop is handled via adjusting the available current, the cable remains locked at all times. If pause/resume is used, it will be unlocked when paused. Presenting the RFID key at any time will unauthorize the charging and the cable will be released.

- Set authentication enable and add an RFID tag in zaptec app
- Set available current to 0 to prevent automatic charge start
- Insert cable. Observe yellow blinking (waiting for authentication)
- Present RFID tag. Observe green blinking
- Set available current to wanted value which will start the charge process
- Issuing "stop_pause_charging" (506) stops charging and puts it into "finished" mode and unlocks the cable.
- Issuing "resume_charging" (507) starts charging again (unless current is set to 0).
- Setting available current to 0 sets stop charging and puts it into "finished" mode, but the cable remains locked

I think this is a very attractive use model. It has the advantage that the RFID authentication ID is tagged with the session. Having two different RFID keys and two EVs, it can be used to identify which car is connected for charging. This allows for gathering statistics on EVs that doesn't have HA integration. It can enable different charging profile (current) in HA depending on the car connected. And so on.

PS! I hope I'm not deteriorating my EV by doing many small charging sessions. This evening I've charged 8 very short sessions with very miniscule amounts of energy in order to all the testing in these recent posts. 😎

**Comment by MarkMikkelsen83** [comment]
Hey - Im trying to set the max limit but can't make it work via an automation - can somebody help here ?

**Comment by hanpq** [comment]
I just want to shim in regarding the use of authorization regarding the following statement;

> Why do users have/want authorization? I think the answer to that question will determine the need for deauthorize or not.

I'm actually surprised to not see almost everyone having authorization on, I thought it was close to a must have. With authorization off anyone can just drive up to the charger and charge for "free"? I guess you can argue it is unlikely but the same goes with the lock on the house door. Unlikely that someone will just walk in, but in case someone tries it will be locked, same goes with the charger.

Based on this thread, most often than not you can assume that if you see a zaptec charger it is probably unlocked and it would probably work to hook up and start charging for free, this surprises me :)

A parameter is of course in what area the charger is located, I guess if your charger is locked up in personal garage or outside of densely populated areas you can argue the risk is close to nil and opt to have charger open. But as soon as your charger is located outside within a reasonably urbanized area I think it is far more common with authorization enabled and definitely something the integration should support.

I also have an Easee charging box at another facility where I have automated authorization when the following three conditions are met;

- Car charger connected: True
- Charge box connected: True
- Car is within geofenced area of charger: True

I would like to make the same automation to auto-authorize using the zaptec-charger, however this is not currently supported so I'm definitely voting for implementing this feature.

**Comment by hanpq** [comment]
I just want to shim in regarding the use of authorization regarding the following statement;

> Why do users have/want authorization? I think the answer to that question will determine the need for deauthorize or not.

I'm actually surprised to not see almost everyone having authorization on, I thought it was close to a must have. With authorization off anyone can just drive up to the charger and charge for "free"? I guess you can argue it is unlikely but the same goes with the lock on the house door. Unlikely that someone will just walk in, but in case someone tries it will be locked, same goes with the charger.

Based on this thread, most often than not you can assume that if you see a zaptec charger it is probably unlocked and it would probably work to hook up and start charging for free, this surprises me :)

A parameter is of course in what area the charger is located, I guess if your charger is locked up in personal garage or outside of densely populated areas you can argue the risk is close to nil and opt to have charger open. But as soon as your charger is located outside within a reasonably urbanized area I think it is far more common with authorization enabled and definitely something the integration should support.

I also have an Easee charging box at another facility where I have automated authorization when the following three conditions are met;

- Car charger connected: True
- Charge box connected: True
- Car is within geofenced area of charger: True

I would like to make the same automation to auto-authorize using the zaptec-charger, however this is not currently supported so I'm definitely voting for implementing this feature.

**Comment by JensDebergh** [comment]
@hanpq This is also a suggestion I made to treat the zaptec by default using authorization. I installed my Zaptec Go and it came with authorization turned on by default. Documentation also hints to going the "security" by default route, which makes sense.

You can easily automate this flow already (I provided the code to work with auth flow already). For now you can add it to your local setup until we get this figured out.

https://github.com/custom-components/zaptec/issues/55#issuecomment-1668607425

My automation:

```
alias: Zaptec - Auto charge electric vehicle
description: >-
  Starts charging the electric car automatically when certain conditions are met
  (off peak, weekend)
trigger:
  - platform: time_pattern
    hours: "22"
  - platform: state
    entity_id:
      - binary_sensor.the_batmobile_charger
    attribute: charging_state
    from: Disconnected
    to: NoPower
    for:
      hours: 0
      minutes: 0
      seconds: 0
condition:
  - condition: and
    conditions:
      - condition: zone
        entity_id: device_tracker.the_batmobile_location_tracker
        zone: zone.home
      - condition: state
        entity_id: binary_sensor.the_batmobile_charger
        attribute: charging_state
        state:
          - NoPower
      - condition: and
        conditions:
          - condition: or
            conditions:
              - condition: time
                after: "22:00:00"
              - condition: state
                entity_id: binary_sensor.workday_sensor
                state: "off"
action:
  - service: zaptec.authorize_charge
    data:
      charger_id: <YOUR CHARGER ID>
mode: single
```
```

**Comment by hanpq** [comment]
Thank you! It worked perfectly! 

Sorry I read your posts but didn't understand I could just update the files locally directly. Thought I somehow had to repackage the integration or something which is way over my head.

**Comment by sveinse** [comment]
@JensDebergh Cool. Now I see why you want authorization. You use it as a combined start of charge and authorize charging operation. If I understand this correctly, you have no other need of authorization than "not everyone else use the charger" and you have trust to HA?

On a side note, I see from the data from Zaptec that when I use RFID tags for authentication, I get info about the source of the authentication. I'm speculating if the `authorizecharge` API calls takes a JSON payload we don't know about and that payload can contain an identifier to what source the authorization came from.

**Comment by JensDebergh** [comment]
> @JensDebergh Cool. Now I see why you want authorization. You use it as a combined start of charge and authorize charging operation. If I understand this correctly, you have no other need of authorization than "not everyone else use the charger" and you have trust to HA?
> 
> On a side note, I see from the data from Zaptec that when I use RFID tags for authentication, I get info about the source of the authentication. I'm speculating if the `authorizecharge` API calls takes a JSON payload we don't know about and that payload can contain an identifier to what source the authorization came from.

Correct! I don't really care who's charging, because its going to be assigned to my company billing anyway. I just pull the charging statistics from the API and reimburse it to my private household from the company.

You can't get the individual charging statistics with the current implementation of zaptec in home assistant because it's only possible to authorize with a single account configured in the integration.

For that scenario it's best to use the RFID tags I guess. My use case was just to plug it in and its allowed to charge based on my conditions.

- During the week after 22:00 (off peak)
- Weekend (off - peak + solar which is the cheapest)

Currently working on a new integration which accounts for peak capacity (the amount of watts you pull from the grid in a timespan of 15 minutes - we get billed extra if it is too high)

So I'm pausing / starting the charging session based on that information. Combining a washing machine + electric vehicle charging is devastating for this metric, so I'm reading out those values and pause the session when it starts to get too high.

**Comment by hanpq** [comment]
> @JensDebergh Cool. Now I see why you want authorization. You use it as a combined start of charge and authorize charging operation. If I understand this correctly, you have no other need of authorization than "not everyone else use the charger" and you have trust to HA?

Exactly, we don't care much who starts the charging session, rather what car should be charged, if HA concludes it is likely our car connected we allow charging, ruling out any charging when we are not home. If guests need to charge we will just authorize manually in the app. Best case would be if the charging standard provided a unique identifier for the car that could automatically be authorized by the charging box. But that is another discussion not related to HA/Zaptec :)

We could of course use a RFID tag however the chargebox is not easily accessable from where the car is parked so currently (as of a few minutes ago) we authorized within the app to start charging. With this automation we just need to plugin the car and HA authorizes the charging session for us.

**Comment by sveinse** [comment]
Thank you @JensDebergh and @hanpq.

In a general usage context, not all cars have HA integration and some doesn't even offer location services. In these use cases authentication can't be automated on the above conditions. This is where RFID _could_ offer an alternative.

It is important to be aware that when the charging is paused, the cable is unlocked and it will lock itself when charging resumes. I hope and assumes that the Zaptec HW logic is such that if the cable is removed when its paused, it will deauthorize the session and thus require a new authentication when charging should resume.

**Comment by hanpq** [comment]
> It is important to be aware that when the charging is paused, the cable is unlocked and it will lock itself when charging resumes. I hope and assumes that the Zaptec HW logic is such that if the cable is removed when its paused, it will deauthorize the session and thus require a new authentication when charging should resume.

I'll see if I can test that, I use permanently locked cable and I need to re-authorize every time the car is disconnected.

**Comment by sveinse** [comment]
@hanpq Permanent locked cables is ok. It remains locked all the time until its unlocked from the app. We should probably add the function to be able to lock and unlock the permanent lock from HA.

**Comment by sveinse** [comment]
@JensDebergh the new PR #56, which is at beta test in my fork at https://github.com/sveinse/zaptec support authorization. The beta is installable (yet I'd be careful on production environments) via HACS. See the URL for instructions if you want to give it a go.

**Comment by sveinse** [comment]
@JensDebergh version 0.7.0 is now released as beta. It includes support for charge authorization. Please test it and give your feedback.

**Comment by JensDebergh** [comment]
@sveinse Sorry for the delay. Got occupied with work. Just updated all my migrations, it's a BIG improvement to what it previously was.

Using the services is just as simple as clicking the right entities instead of having to lookup your zaptec charger ID. 👍
